src/mgr/DaemonServer.cc

```diff
@@ -2733,11 +2731,9 @@ void DaemonServer::handle_conf_change(const ConfigProxy& conf,
 				      const std::set <std::string> &changed)
 {
   // We may be called within lock (via MCommand `config set`) or outwith the
-  // lock (via admin socket `config set`), so handle either case.
-  const bool initially_locked = lock.is_locked_by_me();
-  if (!initially_locked) {
-    lock.Lock();
-  }
+  // lock (via admin socket `config set`).
+  // Since lock is recursive, it should handle either case.
+  std::lock_guard l(lock);
```

Comment:
Instead of introducing a recursive lock we could use the `finisher` queue,  similarly to what `DaemonServer::handle_osd_perf_metric_query_updated` already does.

Reply:
Aah, very nice. Thank you @trociny !